raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1902,7 +1902,11 @@ public Long scheduleAppend(int epoch, List<T> records) {
 
     @Override
     public LeaderAndEpoch leaderAndEpoch() {
-        return quorum.leaderAndEpoch();
+        LeaderAndEpoch leaderAndEpoch = new LeaderAndEpoch(OptionalInt.empty(), -1);
+        if (quorum != null) {
```

Comment:
Hmm... This seems to suggest we are using `leaderAndEpoch` before initialization has completed. I am thinking how we can avoid that. If you look in `KafkaRaftClient.initialize()`, the only reason we need the addresses is to initialize the endpoints:
```java
        Map<Integer, InetSocketAddress> voterAddresses = raftConfig.quorumVoterConnections();
        for (Map.Entry<Integer, InetSocketAddress> voterAddressEntry : voterAddresses.entrySet()) {
            channel.updateEndpoint(voterAddressEntry.getKey(), voterAddressEntry.getValue());
        }
```
In fact, there's no reason why this must be done in `initialize`. It could also be done in `RaftManager.startup`. In fact, if you look at the handling inside `KafkaNetworkChannel`, we already handle the case when the destination address is not defined. So what we could do is something like this:

1. Initialize `controller.quorum.voters` using non-routable addresses. For example, we could use `1@0.0.0.0:0,2@0.0.0.0:0`. 
2. Add a method to `RaftConfig` which parses only the voter ids. This will be used to initialize the quorum state and such in `KafkaRaftClient.initialize`.
3. Move endpoint initialization out of `KafkaRaftClient.initialize` and into `RaftManager`. Basically when the future completes, we can call `KafkaNetworkChannel.updateEndpoint`.

Reply:
I think that makes sense. We can generate a quorum voters string using the number of controllers (which should be fixed and known up-front even in the TestKit) and the non-routable address

Reply:
We don't need to change anything in `RaftConfig.quorumVoterIds` actually if we pass in a connection string that is of the form you specified. It'll parse the voter IDs out as is. So, I think the only change we need is to handle the endpoint update outside in the RaftManager and also update the endpoints in the RaftConfig to keep it consistent

Reply:
Even better. I wasn't sure if `InetSocketAddress` validated that the ip was routable.